trainer.py

```diff
@@ -0,0 +1,102 @@
+import numpy as np
+import torch
+import tqdm
+from torch.nn import CTCLoss
+from torch.optim import Adam
+from torch.utils.tensorboard import SummaryWriter
+
+from dfa.dataset import new_dataloader, get_longest_mel_id
+from dfa.duration_extraction import extract_durations_with_dijkstra
+from dfa.model import Aligner
+from dfa.paths import Paths
+from dfa.text import Tokenizer
+from dfa.utils import to_device
+
+
+class Trainer:
+
+    def __init__(self, paths: Paths) -> None:
+        self.paths = paths
+        self.writer = SummaryWriter(log_dir=paths.checkpoint_dir / 'tensorboard', comment='v1')
+        self.ctc_loss = CTCLoss()
+
+        # Used for generating plots
+        longest_id = get_longest_mel_id(dataset_path=self.paths.data_dir / 'dataset.pkl')
+        self.longest_mel = np.load(str(paths.mel_dir / f'{longest_id}.npy'), allow_pickle=False)
+        self.longest_tokens = np.load(str(paths.token_dir / f'{longest_id}.npy'), allow_pickle=False)
+
+    def train(self, checkpoint: dict, train_params: dict) -> None:
+        lr = train_params['learning_rate']
+        epochs = train_params['epochs']
+        batch_size = train_params['batch_size']
+        ckpt_steps = train_params['checkpoint_steps']
+        plot_steps = train_params['plot_steps']
+
+        config = checkpoint['config']
+        symbols = checkpoint['symbols']
+        tokenizer = Tokenizer(symbols)
+        device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
+
+        model = Aligner.from_checkpoint(checkpoint).to(device)
+        optim = Adam(model.parameters())
+        optim.load_state_dict(checkpoint['optim'])
+
+        for g in optim.param_groups:
+            g['lr'] = lr
```

Comment:
why not optim = optim.Adam(model.parameters(), lr=train_params['learning_rate'])?

Reply:
I think it loads the learning rate from the state dict and we have to override it.